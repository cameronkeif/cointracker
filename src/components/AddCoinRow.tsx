import React, { ChangeEvent, KeyboardEvent } from 'react';
import { green, grey } from '@material-ui/core/colors';
import IconButton from '@material-ui/core/IconButton';
import TextField from '@material-ui/core/TextField';
import AddCircleIcon from '@material-ui/icons/AddCircle';
import Autocomplete, { AutocompleteRenderInputParams } from '@material-ui/lab/Autocomplete';

import { CoinOption } from '../react-app-env';
import coinOptions from '../data/coin-options.json';

export type AddCoinRowProps = {
  onAdd: Function,
  selectedSymbols: Set<string>
};

const AddCoinRow = ({ onAdd, selectedSymbols }: AddCoinRowProps) => {
  const [selectedCoin, setSelectedCoin] = React.useState<null | CoinOption>(null);

  const addCoin = () => {
    if (selectedCoin === null) {
      return;
    }

    onAdd(selectedCoin);

    setSelectedCoin(null);
  };

  const addButtonColor = selectedCoin ? green[500] : grey[500];
  return (
    <div style={{
      display: 'inline-flex', alignItems: 'center', width: 'auto', marginTop: 20,
    }}
    >
      <Autocomplete
        id="combo-box-demo"
        options={coinOptions.filter((coinOption: CoinOption) => !selectedSymbols.has(coinOption.symbol))}
        style={{ width: 300, margin: 'auto', display: 'inline-block' }}
        renderInput={(params: AutocompleteRenderInputParams) => (
          <TextField
            {...params}
            label="Add coin"
            size="small"
            variant="outlined"
            data-id="coin-input"
          />
        )}
        getOptionLabel={(coinOption: CoinOption) => `${coinOption.name} (${coinOption.symbol})`}
        onChange={(
          event: ChangeEvent<{}>,
          newSelectedCoin: { name: string; symbol: string; } | null,
        ) => { setSelectedCoin(newSelectedCoin); }}
        onKeyDown={(event: KeyboardEvent<HTMLDivElement>) => {
          if (event.key === 'Enter') {
            addCoin();
          }
        }}
        value={selectedCoin}
      />
      <IconButton
        disabled={!selectedCoin}
        onClick={addCoin}
        data-id="add-btn"
      >
        <AddCircleIcon
          style={{ color: addButtonColor, cursor: 'pointer' }}
          fontSize="large"
        />
      </IconButton>
    </div>
  );
};

export default AddCoinRow;
