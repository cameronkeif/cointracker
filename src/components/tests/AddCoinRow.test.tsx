import React from 'react';
import { shallow, mount } from 'enzyme';
import IconButton from '@material-ui/core/IconButton';
import Autocomplete from '@material-ui/lab/Autocomplete';

import AddCoinRow, { AddCoinRowProps } from '../AddCoinRow';

const render = (renderFunction: Function, props: AddCoinRowProps) => shallow(<AddCoinRow {...props} />);

const selectedSymbols = new Set(['BTC', 'ETH']);

describe('AddCoinRow', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly', () => {
    expect(render(shallow, { onAdd: jest.fn(), selectedSymbols: new Set() })).toMatchSnapshot();
  });

  it('renders correctly (with selectedSymbols populated)', () => {
    expect(render(shallow, { onAdd: jest.fn(), selectedSymbols })).toMatchSnapshot();
  });

  it('filters out options matching the provided selectedSymbols', () => {
    const component = render(mount, { onAdd: jest.fn(), selectedSymbols });
    component.find(Autocomplete).props().options
      .forEach((option) => {
        expect(option.symbol).not.toBe('BTC');
        expect(option.symbol).not.toBe('ETH');
      });
  });

  it('Disables the add button if selectedCoin is null', () => {
    const addButton = render(mount, { onAdd: jest.fn(), selectedSymbols }).find(IconButton);
    expect(addButton.props().disabled).toBe(true);
  });

  it('Calls onAdd when the add button is clicked or the enter button is pressed', () => {
    jest
      .spyOn(React, 'useState')
      .mockImplementationOnce(() => React.useState({ selectedCoin: { name: 'Bitcoin', smybol: 'BTC' } }));

    const onAdd = jest.fn();

    // Test click
    let component = render(mount, { onAdd, selectedSymbols });
    const addButton = component.find(IconButton);

    expect(addButton.props().disabled).toBe(false);
    expect(onAdd.mock.calls.length).toBe(0);

    addButton.simulate('click');
    expect(onAdd.mock.calls.length).toBe(1);

    // Re-render and test "enter keydown"
    jest
      .spyOn(React, 'useState')
      .mockImplementationOnce(() => React.useState({ selectedCoin: { name: 'Bitcoin', smybol: 'BTC' } }));

    component = render(mount, { onAdd, selectedSymbols });
    const autoComplete = component.find(Autocomplete);

    autoComplete.simulate('keydown', { key: 'Enter' });
    expect(onAdd.mock.calls.length).toBe(2);
  });
});
