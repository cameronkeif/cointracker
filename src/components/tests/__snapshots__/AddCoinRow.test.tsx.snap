// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddCoinRow renders correctly (with selectedSymbols populated) 1`] = `ShallowWrapper {}`;

exports[`AddCoinRow renders correctly 1`] = `ShallowWrapper {}`;
